build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/core/HeadlessBuilder.java

```diff
@@ -1044,4 +1048,53 @@ protected void resetToolOptions(IConfiguration configuration) throws BuildExcept
 	@Override
 	public void stop() {
 	}
+
+	/*
+	 * Sort projects so that referenced projects are built first.
+	 * E.g. for references A -> B -> C (A references B, B references C),
+	 * we want the order to be: C, B, A
+	 * For projects that are on a cycle or have no reference relationship, sort by project name.
+	 */
+	private static List<IProject> sortProjects(Set<IProject> projs) throws CoreException {
+		List<IProject> projects = new ArrayList<>();
+		projects.addAll(projs);
+		/*
+		 * Gather all references of each project, including indirect ones.
+		 * E.g. for A -> B -> C, the resulting map will contain
+		 * A -> {B, C}
+		 * B -> { C }
+		 */
+		Map<IProject, Set<IProject>> references = new HashMap<>();
+		for (IProject project : projects) {
+			Set<IProject> referencesOfProject = new HashSet<>();
+			references.put(project, referencesOfProject);
+			Deque<IProject> queue = new LinkedList<>();
+			queue.add(project);
+			while (!queue.isEmpty()) {
+				IProject p = queue.removeFirst();
+				referencesOfProject.add(p);
```

Comment:
Does this code handle circular dependencies? On a quick look it seems to me this is an infinite loop if A -> B -> A is in the reference list. The sorting code's comment indicates you considered a cycle, so if this code handles the cycle that is good.

Perhaps this needs to check if `referencesOfProject` contains `p` first?

Reply:
Ah I've forgotten a visited set here I'll add one tomorrow, hopefully with tests for such cases too. Thanks!